botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1123 @@
+VALID_HOST_LABEL_RE = re.compile(
+    r"^((?!-)[a-zA-Z\d-]{1,63}(?<!-))*((?!-)[a-zA-Z\d-]{1,63}(?<!-))$",
+)
+GET_ATTR_RE = re.compile(r"(\w+)\[(\d+)\]")
+CACHE_SIZE = 100
+
+
+class BaseRule:
+    """A rule within a rule set. All rules contain a conditions property,
+    which can be empty, a partitions JSON and documentation about the rule.
+    """
+
+    def __init__(self, conditions, partitions, documentation=None):
+
+        self.conditions = conditions
+        self.partitions = partitions
+        self.documentation = documentation
+
+    def evaluate(self, scope_vars):
+        raise NotImplementedError()
+
+    def is_func(self, argument):
+        """Determine if an object is a function object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a function object.
+
+        :returns: A bool indicating if the input is a function object.
+        """
+        return isinstance(argument, dict) and "fn" in argument
+
+    def is_ref(self, argument):
+        """Determine if an object is a reference object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a reference object.
+
+        :returns: A bool indicating if the input is a reference object.
+        """
+        return isinstance(argument, dict) and "ref" in argument
+
+    def is_template(self, argument):
+        """Determine if an object contains a template string.
+
+        :type argument: Any
+        :param argument: An object that may or may not contain a template string.
+
+        :returns: A bool indicating if the input contains a template string.
+        """
+        return (
+            isinstance(argument, str)
+            and TEMPLATE_STRING_RE.search(argument) is not None
+        )
+
+    def resolve_template_string(self, value, scope_vars):
+        """Given a templated string, parse out the template nested between curly
+        braces, and return the nested data from the rule's scoped variables. If the
+        string contains double curly braces such as `{{foo}}` they are treated as
+        the string literal `{foo}` and are not formatted.
+
+        :type value: str
+        :param value: A template string such as "https://{Bucket}.ec2.{url#authority}".
+
+        :type scope_vars: dict
+        :param scope_vars: A dictionary combining input parameters and variables that
+        have been scoped throughout the rule's evaluation.
+
+        :returns: A fully resolved string such as "https://us-west-2.ec2.amazonaws.com".
+        """
+        string_formatter = Formatter()
+        result = ""
+        for literal, reference, _, _ in string_formatter.parse(value):
+            if reference is not None:
+                template_value = scope_vars
+                template_params = reference.split("#")
+                for param in template_params:
+                    template_value = template_value[param]
+                result += f"{literal}{template_value}"
```

Comment:
Could do this as just f-string instead of f-string and string addition:

```suggestion
                result = f"{result}{literal}{template_value}"
```

Reply:
why is that better? i'd much rather append to string using += instead of writing the variable for the string twice

Reply:
Good question, and actually I have to take my comment back. I was under the common (I think?) impression that string concatenation with the `+` operator is always slower than all alternatives and that f-strings are generally the fastest. It turns out that `+=` is one of the few exceptions to that rule. This blog post (by an author who also got surprised by this) educated me and explains it well: https://blog.ganssle.io/articles/2019/11/string-concat.html

Reply:
oh interesting i never knew. i figured under the hood it was doing the same thing `+=` was just shorthand.